### PR TITLE
build(python): Update to PyO3 to 0.18.0

### PR DESCRIPTION
**Pull request by jjerphan**
Follow-up of #4794

### CODE REVIEW

**Comment by jjerphan** [comment]
I still have tests failure because NumPy arrays containing strings cannot be converted to `pl.Series`. I am currently trying to understand why this is the case.

**Comment by ritchie46** [comment]
I believe it is related to https://github.com/PyO3/pyo3/issues/2615

This blocked previous upgrade.

**Comment by ritchie46** [comment]
@jjerphan maybe we can add a new constructor specialized for this case? 

Or we could ask upstream what is best way to do this in latest release?

**Comment by jjerphan** [comment]
I guess we can ask upstream and in the mean time have a new constructor which has been specialized for this case and a comment indicating that such a constructor is temporary.

What do you think of this approach?

**Comment by ritchie46** [comment]
Yes, I think that's a good idea.

**Comment by jjerphan** [comment]
I am busy now, but I should be able to have a look at it tomorrow.

File: examples/python_rust_compiled_function/src/ffi.rs [hunk 1]
```diff
@@ -71,7 +71,7 @@ pub fn rust_series_to_py_series(series: &Series) -> PyResult<PyObject> {
     let array = series.to_arrow(0);
 
     // acquire the gil
-    let gil = Python::acquire_gil();
+    let gil = Python::with_gil();
```
Comment (adamreichold) [inline examples/python_rust_compiled_function/src/ffi.rs:74]: This does not look correct. Is this example built in your CI? [`with_gil`](https://docs.rs/pyo3/latest/pyo3/marker/struct.Python.html#method.with_gil) takes a closure which gets the GIL token passed as an argument.

**Comment by adamreichold** [comment]
Could someone point me to where the entry point for those failing tests is so that I can check which downcast/extraction is failing exactly? Thank you!

File: examples/python_rust_compiled_function/src/ffi.rs (hunk 1 above)
Comment (jjerphan) [reply examples/python_rust_compiled_function/src/ffi.rs:74]: I made this change because I got a deprecation warning locally.

**Comment by ritchie46** [comment]
@adamreichold I can come back to you in an hour. 👍

**Comment by adamreichold** [comment]
> Could someone point me to where the entry point for those failing tests is so that I can check which downcast/extraction is failing exactly? Thank you!

So by digging and observing that this seems to affect only NumPy arrays of strings, I suspect that this comes from `PySeries::new_str` which extracts `Wrap<Utf8Chunked>`, right? (rust-numpy does not support the `<U*` dtypes yet.)

File: examples/python_rust_compiled_function/src/ffi.rs (hunk 1 above)
Comment (adamreichold) [reply examples/python_rust_compiled_function/src/ffi.rs:74]: The deprecation is intentional, but `Python::with_gil` just works differently. This PR already contains other places where that change is made correctly, so you should be able to copy from there.

**Comment by adamreichold** [comment]
> > Could someone point me to where the entry point for those failing tests is so that I can check which downcast/extraction is failing exactly? Thank you!
> 
> So by digging and observing that this seems to affect only NumPy arrays of strings, I suspect that this comes from `PySeries::new_str` which extracts `Wrap<Utf8Chunked>`, right? (rust-numpy does not support the `<U*` dtypes yet.)

I think I found the issue, your various impls like

```rust
impl<'a, T> FromPyObject<'a> for Wrap<ChunkedArray<T>>
```

use a helper function

```rust
pub(crate) fn get_pyseq(obj: &PyAny) -> PyResult<(&PySequence, usize)> {
    let seq = <PySequence as PyTryFrom>::try_from(obj)?;
    let len = seq.len()?;
    Ok((seq, len))
}
```

and this will indeed fail with NumPy arrays as we do not downcast those to `PySequence` any more after https://github.com/PyO3/pyo3/pull/2477 because it does not really support the sequence API (even though they do pass `PySequence_Check`).

But I also think that your code does not actually use the sequence interface as both `.len()` and `.iter()` work on any `PyAny`. So dropping `get_pyseq` together with mechanical changes like

```diff
diff --git a/py-polars/src/conversion.rs b/py-polars/src/conversion.rs
index 463b65371..575d69fbd 100644
--- a/py-polars/src/conversion.rs
+++ b/py-polars/src/conversion.rs
@@ -125,10 +125,10 @@ impl<'a> FromPyObject<'a> for Wrap<BooleanChunked> {
 
 impl<'a> FromPyObject<'a> for Wrap<Utf8Chunked> {
     fn extract(obj: &'a PyAny) -> PyResult<Self> {
-        let (seq, len) = get_pyseq(obj)?;
+        let len = obj.len()?;
         let mut builder = Utf8ChunkedBuilder::new("", len, len * 25);
 
-        for res in seq.iter()? {
+        for res in obj.iter()? {
             let item = res?;
             match item.extract::<&str>() {
                 Ok(val) => builder.append_value(val),
```

should work.

File: examples/python_rust_compiled_function/src/ffi.rs (hunk 1 above)
Comment (ritchie46) [reply examples/python_rust_compiled_function/src/ffi.rs:74]: @jjerphan this should be:

```
pub fn rust_series_to_py_series(series: &Series) -> PyResult<PyObject> {
    // ensure we have a single chunk
    let series = series.rechunk();
    let array = series.to_arrow(0);

    Python::with_gil(|py| {
        // import pyarrow
        let pyarrow = py.import("pyarrow")?;

        // pyarrow array
        let pyarrow_array = to_py_array(py, pyarrow, array)?;

        // import polars
        let polars = py.import("polars")?;
        let out = polars.call_method1("from_arrow", (pyarrow_array,))?;
        Ok(out.to_object(py))
    })
}
```

**Comment by ritchie46** [comment]
Thanks a lot @adamreichold. I believe that was the issue indeed. Will come back with a confirmation. I

**Comment by jjerphan** [comment]
It looks like default values for all optional parameters must be provided as "required arguments after an `Option<_>` argument are ambiguous and being phased out".

**Comment by ritchie46** [comment]
> It looks like default values for all optional parameters must be provided as "required arguments after an `Option<_>` argument are ambiguous and being phased out".

Yes, everything with an `Option` now needs a attribute:

```
#[pyo3(signature=(a, b))]
fn do(a: u64, b: Option<u64>) {}  
```

**Comment by jjerphan** [comment]
Is there a simple way to infer default values for signatures' arguments?

**Comment by ritchie46** [comment]
In polars rust side all arguments are always required. So signature should just name all arguments.

**Comment by jjerphan** [comment]
At the moment, I get this kind of warning:

```
warning: use of deprecated constant `pyo3::impl_::deprecations::REQUIRED_ARGUMENT_AFTER_OPTION`: required arguments after an `Option<_>` argument are ambiguous and being phased out
         = help: add a `#[pyo3(signature)]` annotation on this function to unambiguously specify the default values for all optional parameters
  --> src/batched_csv.rs:25:9
   |
25 |         chunk_size: usize,
   |         ^^^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default
```

I've tried to only specify the list of arguments' names, but I get other errors.

It seems that default values for required arguments must be provided.

**Comment by ritchie46** [comment]
Did you add `#[pyo3(signature...`?

See: https://pyo3.rs/v0.18.0/migration#required-arguments-after-option_-arguments-will-no-longer-be-automatically-inferred

**Comment by jjerphan** [comment]
Everything should be up to date.

Please, let me know if `'#[pyo3(signature = ...)]'` macros are properly formatted.

File: polars/polars-lazy/Cargo.toml [hunk 2]
```diff
@@ -24,7 +24,7 @@ polars-pipe = { version = "0.26.1", path = "./polars-pipe", optional = true }
 polars-plan = { version = "0.26.1", path = "./polars-plan" }
 polars-time = { version = "0.26.1", path = "../polars-time", optional = true }
 polars-utils = { version = "0.26.1", path = "../polars-utils" }
-pyo3 = { version = "0.16", optional = true }
+pyo3 = { version = "0.18", optional = true }
```
Comment (jjerphan) [inline polars/polars-lazy/Cargo.toml:27]: Must this include the bug-fix number? I.e. must it be:
```suggestion
pyo3 = { version = "0.18.0", optional = true }
```
?

File: polars/polars-lazy/polars-plan/Cargo.toml [hunk 3]
```diff
@@ -17,7 +17,7 @@ polars-io = { version = "0.26.1", path = "../../polars-io", features = ["lazy",
 polars-ops = { version = "0.26.1", path = "../../polars-ops", default-features = false }
 polars-time = { version = "0.26.1", path = "../../polars-time", optional = true }
 polars-utils = { version = "0.26.1", path = "../../polars-utils" }
-pyo3 = { version = "0.16", optional = true }
+pyo3 = { version = "0.18", optional = true }
```
Comment (jjerphan) [inline polars/polars-lazy/polars-plan/Cargo.toml:20]: Similarly: must this include the bug-fix number? I.e. must it be:
```suggestion
pyo3 = { version = "0.18.0", optional = true }
```
?

File: py-polars/Cargo.toml [hunk 4]
```diff
@@ -19,11 +19,11 @@ bincode = "1.3"
 # todo: unfix when compilation problem is solved
 libc = "0.2"
 ndarray = "0.15"
-numpy = "0.16"
+numpy = "0.18"
```
Comment (jjerphan) [inline py-polars/Cargo.toml:22]: Similarly: must this include the bug-fix number? I.e. must it be:
```suggestion
numpy = "0.18.0"
```

File: py-polars/src/lazy/dataframe.rs [hunk 5]
```diff
@@ -592,6 +605,7 @@ impl PyLazyFrame {
 
     #[allow(clippy::too_many_arguments)]
     #[cfg(feature = "asof_join")]
+    #[pyo3(signature = (other, left_on, right_on, left_by, right_by, allow_parallel, force_parallel, suffix, strategy, tolerance, tolerance_str))]
```
Comment (jjerphan) [inline py-polars/src/lazy/dataframe.rs:608]: Must such long macro lines be reformatted?

File: py-polars/src/conversion.rs [hunk 6]
```diff
@@ -125,10 +125,10 @@ impl<'a> FromPyObject<'a> for Wrap<BooleanChunked> {
 
 impl<'a> FromPyObject<'a> for Wrap<Utf8Chunked> {
     fn extract(obj: &'a PyAny) -> PyResult<Self> {
-        let (seq, len) = get_pyseq(obj)?;
+        let len = obj.len()?;
```
Comment (jjerphan) [inline py-polars/src/conversion.rs:128]:
```suggestion
        # TODO: extend `get_pyseq` for this case and replace
        # `obj.len()?` here by `get_pyseq(obj)?`.
        let len = obj.len()?;
```

File: py-polars/polars/internals/lazyframe/frame.py [hunk 7]
```diff
@@ -230,6 +230,7 @@ def _scan_parquet(
             rechunk,
             _prepare_row_count_args(row_count_name, row_count_offset),
             low_memory,
+            cloud_options=storage_options,
```
Comment (jjerphan) [inline py-polars/polars/internals/lazyframe/frame.py:233]: Is this change wrong, or is there a naming discrepancy here?

File: py-polars/src/npy.rs [hunk 8]
```diff
@@ -32,7 +32,7 @@ pub unsafe fn aligned_array<T: Element + NativeType>(
     let len = buf.len();
     let buffer_ptr = buf.as_mut_ptr();
 
-    let dims = [len].into_dimension();
+    let mut dims = [len].into_dimension();
```
Comment (jjerphan) [inline py-polars/src/npy.rs:35]: Do we need to motivate the `mut` qualifier here?

**Review by jjerphan — COMMENTED** [review]
Here are some questions regarding some of the proposed changes.

File: py-polars/src/conversion.rs (hunk 6 above)
Comment (adamreichold) [reply py-polars/src/conversion.rs:128]: I would rather suggest to drop `get_pyseq` entirely. It is not clear which purposes it serves if the `PySequece` API is not used.

File: polars/polars-lazy/Cargo.toml (hunk 2 above)
Comment (ritchie46) [reply polars/polars-lazy/Cargo.toml:27]: I prefer 0.18 as this makes it more clear that we accept future patches.

File: py-polars/src/lazy/dataframe.rs (hunk 5 above)
Comment (ritchie46) [reply py-polars/src/lazy/dataframe.rs:608]: If cargo fmt doesn't do it, it seems fine. :+1:

File: py-polars/src/npy.rs (hunk 8 above)
Comment (ritchie46) [reply py-polars/src/npy.rs:35]: Why do we need a mut here?

Comment (jjerphan) [reply py-polars/src/npy.rs:35]: This was suggested by the compiler.

```
error[E0596]: cannot borrow `dims` as mutable, as it is not declared as mutable
  --> src/npy.rs:43:9
   |
43 |         dims.as_dims_ptr(),
   |         ^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
35 |     let mut dims = [len].into_dimension();
   |         +++
```

I guess it better be `const`-qualified instead (I am trying to see how I can type it).

Comment (ritchie46) [reply py-polars/src/npy.rs:35]: Alright, I think the mut is fine. The signatures would be explanatory enough upon inspection.

**Comment by ritchie46** [comment]
Thanks a lot @jjerphan and @adamreichold for the help.

Let me know when it is good to go.

**Comment by jjerphan** [comment]
Thank you for your guidance, @ritchie46.

I think this is mergeable.

**Comment by jjerphan** [comment]
Thank you for your guidance, @ritchie46 and @adamreichold.

I think this is mergeable.

**Pull request merged**